server/src/com/thoughtworks/go/server/messaging/plugin/StageStatusPluginNotifier.java

```diff
@@ -16,49 +16,258 @@
 
 package com.thoughtworks.go.server.messaging.plugin;
 
-import com.thoughtworks.go.domain.Stage;
+import com.thoughtworks.go.config.materials.PackageMaterial;
+import com.thoughtworks.go.config.materials.PluggableSCMMaterial;
+import com.thoughtworks.go.config.materials.dependency.DependencyMaterial;
+import com.thoughtworks.go.config.materials.git.GitMaterial;
+import com.thoughtworks.go.config.materials.mercurial.HgMaterial;
+import com.thoughtworks.go.config.materials.perforce.P4Material;
+import com.thoughtworks.go.config.materials.svn.SvnMaterial;
+import com.thoughtworks.go.config.materials.tfs.TfsMaterial;
+import com.thoughtworks.go.domain.*;
+import com.thoughtworks.go.domain.config.Configuration;
+import com.thoughtworks.go.domain.config.ConfigurationProperty;
+import com.thoughtworks.go.domain.materials.Material;
+import com.thoughtworks.go.domain.materials.Modification;
+import com.thoughtworks.go.domain.packagerepository.PackageDefinition;
+import com.thoughtworks.go.domain.packagerepository.PackageRepository;
+import com.thoughtworks.go.domain.scm.SCM;
 import com.thoughtworks.go.plugin.access.notification.NotificationExtension;
+import com.thoughtworks.go.plugin.access.notification.NotificationPluginRegistry;
 import com.thoughtworks.go.server.domain.StageStatusListener;
+import com.thoughtworks.go.server.service.PipelineService;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
 
 import java.sql.Timestamp;
 import java.text.SimpleDateFormat;
-import java.util.LinkedHashMap;
-import java.util.Map;
+import java.util.*;
 
 @Component
 public class StageStatusPluginNotifier implements StageStatusListener {
     public static final String DATE_PATTERN = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
 
+    private NotificationPluginRegistry notificationPluginRegistry;
+    private PipelineService pipelineService;
     private PluginNotificationQueue pluginNotificationQueue;
 
     @Autowired
-    public StageStatusPluginNotifier(PluginNotificationQueue pluginNotificationQueue) {
+    public StageStatusPluginNotifier(NotificationPluginRegistry notificationPluginRegistry, PipelineService pipelineService, PluginNotificationQueue pluginNotificationQueue) {
+        this.notificationPluginRegistry = notificationPluginRegistry;
+        this.pipelineService = pipelineService;
         this.pluginNotificationQueue = pluginNotificationQueue;
     }
 
     @Override
     public void stageStatusChanged(final Stage stage) {
-        Map data = createRequestDataMap(stage);
+        if (notificationPluginRegistry.isAnyPluginInterestedIn(NotificationExtension.STAGE_STATUS_CHANGE_NOTIFICATION)) {
```

Comment:
Also, i have added a check to make sure all the computation happens if someone is interested (a notification plugin exists)